### PR TITLE
[DispatchCreation] Fold collapse(expand) unit dims

### PR DESCRIPTION
**Pull request by IanWood1**
Folds unit dimensions that are created by an expand shape then immediately collapsed. This becomes problematic because during `BubbleUpExpandShapes` these unit dims will get propagated to the surrounding compute ops.

I previously attempted to land this on https://github.com/iree-org/iree/pull/19357

2/2 fixes for #19832

### CODE REVIEW

File: compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp [hunk 1]
```diff
@@ -113,6 +113,163 @@ struct FoldAttentionMaskUnitDim final
...
+
+    const SmallVector<ReassociationIndices, 4> collapseReassoc =
+        collapseOp.getReassociationIndices();
+    ArrayRef<int64_t> interShape = expandOp.getType().getShape();
+    ArrayRef<int64_t> outShape = collapseOp.getType().getShape();
+    SmallVector<int64_t> interToOutMap(expandOp.getType().getRank());
+
+    // Step 1: construct a set of dimensions (with respect to the intermediate
+    // shape) that are unit length and get collapsed away by the final
+    // `collapse_shape` op.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp]: I think the use of backtick symbol (\`) is starting from doxygen tool, which is a render tool for LLVM docs. Most of my usage is for highlighting code elements, i.e., the actual variable in the code snippet. Today, I believe some people like the style because of markdown. (not saying that there is or is not a tool for integrating such stuff using markdown or not.) However, I'd prefer `collapseOp` here because it maps to the original tensor::CollapseShapeOp directly.

File: compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp [hunk 2]
```diff
@@ -113,6 +113,163 @@ struct FoldAttentionMaskUnitDim final
...
+        // kept (prefer the first).
+        if (outShape[outDim] == 1 && innerIdx != 0) {
+          continue;
+        }
+        toDrop.insert(inDim);
+      }
+    }
+
+    // Step 2: remove dimensions from `toDrop` that weren't introduced by the
+    // `expand_shape` op.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp]: same nit: I'd write `expandOp`

File: compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp [hunk 3]
```diff
@@ -113,6 +113,163 @@ struct FoldAttentionMaskUnitDim final
...
+          if (!indices.empty()) {
+            reassoc.push_back(std::move(indices));
+            return true;
+          }
+          return false;
+        };
+
+    auto isIdentityReassociation = [](ArrayRef<ReassociationIndices> reassoc) {
+      return llvm::all_of(reassoc,
+                          [](auto &indices) { return indices.size() == 1; });
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp]: Can this be `llvm::all_of(reassoc, llvm::hasSingleElement)`? btw, maybe move it right before where it is used, i.e., step 6. This way, they can be in the same screen when people browse the code. I think it is one of intentions that makes it lambda, but not a static method?

File: compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp [hunk 4]
```diff
@@ -113,6 +113,163 @@ struct FoldAttentionMaskUnitDim final
...
+    newOutputSizes.reserve(interShape.size() - toDrop.size());
+    SmallVector<OpFoldResult> outputSizes = getMixedValues(
+        expandOp.getStaticOutputShape(), expandOp.getOutputShape(), rewriter);
+    for (auto [idx, ofr] : llvm::enumerate(outputSizes)) {
+      if (!toDrop.contains(idx)) {
+        std::optional<int64_t> staticDim = getConstantIntValue(ofr);
+        newInterShape.push_back(staticDim.value_or(ShapedType::kDynamic));
+        newOutputSizes.push_back(ofr);
+      }
+    }
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp]: Output is a little ambiguous to me because it could be either the output of the expand shape op or collapse shape op. Maybe rename it to `origInterSizes` and `newInterSizes`?

File: compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp [hunk 5]
```diff
@@ -113,6 +113,163 @@ struct FoldAttentionMaskUnitDim final
...
+            if (!toDrop.contains(idx)) {
+              indices.push_back(dim++);
+            }
+          }
+          if (!indices.empty()) {
+            reassoc.push_back(std::move(indices));
+            return true;
+          }
+          return false;
+        };
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp]: optional nit: I'd write down `-> bool` instead of automatic return type deduction because you explicitly say that in the comment. I think automatic return type deduction is useful when the return type may change and we don't really care about the type for maintenance.

nit 2: use `auto` for the line 206, where `llvm::seq` already spells the type.

optional nit: I'd swap the order of `count` and `origStart` in the function arguments because it is clearer that `count` is related to `origStart`. I had such question when I skimmed the code.

File: compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp [hunk 6]
```diff
@@ -113,6 +113,163 @@ struct FoldAttentionMaskUnitDim final
...
+/// %0 = expand_shape ... tensor<3x3x10xf32> into tensor<3x3x5x1x2xf32>
+/// %1 = collapse_shape %1 ... tensor<3x3x5x1x2xf32> into tensor<9x5x2xf32>
+/// ```
+///
+/// simplifies to
+/// ```
+/// %0 = expand_shape ... tensor<3x3x10xf32> into tensor<3x3x5x2xf32>
+/// %1 = collapse_shape %1 ... tensor<3x3x5x2xf32> into tensor<9x5x2xf32>
+/// ```
+struct DropUnitDimsFromCollapseOfExpand
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp:133]: We should add a document about the first unit dim is kept if all the dimensions being collapsed are unit dims.

File: compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp [hunk 7]
```diff
@@ -113,6 +113,163 @@ struct FoldAttentionMaskUnitDim final
...
+                                           expandReassoc[dim].size(),
+                                           expandReassoc[dim].front());
+      if (changed) {
+        expandedDim += newExpandReassoc.back().size();
+      }
+    }
+
+    // Step 6: construct the new `expand_shape` and `collapse_shape` ops.
+    // Note: we must handle the cases where the expand/collapse is no longer
+    // needed.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp]: why? Is it an invalid operation if identity reassociation map is used?

**Review by hanhanW — COMMENTED** [review]
Overall looks good, I'd drop "step numbers" in comments because it likely will end up wrong when code is moved around or when code is inserted/removed.

**Review by MaheshRavishankar — COMMENTED** [review]
I'll let @hanhanW review (and I'll review it tomorrow but no need to wait for me). Please add a TODO as to teach the upstream pattern to handle this as well

File: compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp (hunk 7 above)
Comment (IanWood1) [reply compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp]: Yep, it's invalid. I added a comment.

File: compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp (hunk 6 above)
Comment (IanWood1) [reply compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp:133]: I think this follows from only folding the intermediate dims. For example, the test `collapse_of_expand_4` doesn't remove any unit dims that are on the source of the `expand_shape` or the result of the `collapse_shape` but does remove all of the ones introduced in the intermediate shape.

Comment (hanhanW) [reply compiler/src/iree/compiler/DispatchCreation/FoldUnitExtentDims.cpp:133]: Yes, I understood that, I was mainly talking the implementation comment in line 162-163. If the expanding group has more than one unit dims and the final size is one, the first dimension is preserved.

EDIT: After thinking more, I think my comment is not a concern. It does not matter in the final output, and it is just an implementation detail.

File: compiler/src/iree/compiler/DispatchCreation/test/fold_unit_dims.mlir [hunk 8]
```diff
@@ -202,3 +202,94 @@ util.func public @attention_mask_single_m_dim(%arg0 : tensor<32x1x128xf16>, %arg
...
+  %collapsed = tensor.collapse_shape %expanded [[0], [1, 2, 3]] : tensor<?x1x1x?xf16> into tensor<?x?xf16>
+  util.return %collapsed : tensor<?x?xf16>
+}
+// CHECK-LABEL: util.func public @collapse_of_expand_3
+//  CHECK-SAME:   %[[ARG0:.+]]: tensor<?x?xf16>
+//       CHECK:   util.return %[[ARG0]] : tensor<?x?xf16>
+
+// -----
+
+util.func @collapse_of_expand_4(%arg0: tensor<1x1xf16>, %arg1: index, %arg2: index) -> tensor<1xf16> {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/test/fold_unit_dims.mlir]: nit: maybe rename to all_unit_dims_in_expand_groups

File: compiler/src/iree/compiler/DispatchCreation/test/fold_unit_dims.mlir [hunk 9]
```diff
@@ -202,3 +202,94 @@ util.func public @attention_mask_single_m_dim(%arg0 : tensor<32x1x128xf16>, %arg
...
+}
+// CHECK-LABEL: util.func public @collapse_of_expand_2
+//  CHECK-SAME:   %[[ARG0:.+]]: tensor<?x1xf16>, %[[ARG1:.+]]: index
+//       CHECK:   %[[EXPAND:.+]] = tensor.expand_shape %[[ARG0]]
+//  CHECK-SAME:     tensor<?x1xf16> into tensor<4x?x1xf16>
+//       CHECK:   util.return %[[EXPAND]] : tensor<4x?x1xf16>
+
+// -----
+
+util.func @collapse_of_expand_3(%arg0: tensor<?x?xf16>, %arg1: index, %arg2: index) -> tensor<?x?xf16> {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/DispatchCreation/test/fold_unit_dims.mlir]: maybe rename to `..._fully_dynamic`

**Review by hanhanW — APPROVED** [review]
Just have few suggestions for the function names.
